Use `pkg/util/pointer` functions instead of self-written versions

justaugustus opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Replaces instances of module-written `int(32|64)? --> *int(32|64)?` functions with functions from k8s.io/kubernetes/pkg/util/pointer

**Special notes for your reviewer**:

Here's the grep used, based on the comments in:
* https://github.com/kubernetes/kubernetes/pull/59924#issuecomment-366119396
* https://github.com/kubernetes/kubernetes/issues/59971#issue-297766556

```bash
$ git grep -E 'func\ [^ (]+\([^ ]+\ int(32|64)?\)\ \*int(32|64)?' !(vendor|staging) | grep -v pkg/util/pointer

pkg/apis/apps/v1/defaults_test.go:func newInt32(val int32) *int32 {
pkg/apis/apps/v1beta1/defaults_test.go:func newInt32(val int32) *int32 {
pkg/apis/apps/v1beta2/defaults_test.go:func newInt32(val int32) *int32 {
pkg/apis/autoscaling/v1/defaults_test.go:func newInt32(val int32) *int32 {
pkg/apis/autoscaling/v2beta1/defaults_test.go:func newInt32(val int32) *int32 {
pkg/apis/autoscaling/validation/validation_test.go:func newInt32(val int32) *int32 {
pkg/apis/batch/v1/defaults_test.go:func newInt32(val int32) *int32 {
pkg/apis/batch/v1beta1/defaults_test.go:func newInt32(val int32) *int32 {
pkg/apis/core/v1/defaults_test.go:func newInt(val int32) *int32 {
pkg/apis/core/validation/validation_test.go:func newInt32(val int) *int32 {
pkg/apis/extensions/v1beta1/defaults_test.go:func newInt32(val int32) *int32 {
pkg/controller/deployment/sync_test.go:func intOrStrP(val int) *intstr.IntOrString {
pkg/kubectl/autoscale_test.go:func newInt32(value int) *int32 {
plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:func userIDPtr(i int) *int64 {
plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:func groupIDPtr(i int) *int64 {
test/e2e/apps/deployment.go:func intOrStrP(num int) *intstr.IntOrString {
test/e2e/auth/pod_security_policy.go:func intPtr(i int64) *int64 {
test/integration/deployment/util.go:func intOrStrP(num int) *intstr.IntOrString {
```

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```
/kind cleanup
/cc @php-coder 
/assign @tallclair

justaugustus committed:
pkg/util/pointer: Update `int` pointer functions

* Implement `Int64Ptr` function
* Replace per module functions of `int(32|64)?` --> `*int(32|64)?`
* Update bazel rules

justaugustus commented:
/assign @smarterclayton

php-coder commented on pkg/util/pointer/pointer.go:
```diff
@@ -64,6 +64,12 @@ func IntToInt32Ptr(i int) *int32 {
 }
 
+// IntToInt64Ptr takes an int and returns a pointer to an int64
+func IntToInt64Ptr(i int) *int64 {
```
I doubt that we need this function. Looks like we can use `utilpointer.Int64Ptr()` here because we always specify values explicitly and we can guarantee that these numbers fall within int64 boundaries.

php-coder replied:
Could you try/check this idea and report back whether it works or not?

php-coder commented on pkg/util/pointer/pointer.go:
```diff
@@ -52,6 +52,12 @@ func Int32Ptr(i int32) *int32 {
 }
 
+// IntToInt32Ptr takes an int and returns a pointer to an int32
+func IntToInt32Ptr(i int) *int32 {
```
Consider using `Int32Ptr()` instead of introducing a new function.

php-coder commented:
I'm not sure that it's required but it's always better to put all the autogenerated changes into the single commit. In this case, it would be 2-3 commits with code improvements and one with all autogenerated stuff.

php-coder reviewed (changes requested):
Thank you for another PR! I left a couple comments above.

justaugustus commented:
@php-coder Thanks for the quick review!

I added each of the functions and updated bazel as single commits so that I could easily pluck them out in case we didn't want to use certain functions.

As for introducing the `IntToInt[32|64]` functions, there are a few instances where there are in-module functions to convert from `int` to `int32|int64`.

You can see them if you run this:
```bash
$ git grep -E 'func\ [^ (]+\([^ ]+\ int\)\ \*int(32|64)?' !(vendor|staging) | grep -v pkg/util/pointer

pkg/apis/core/validation/validation_test.go:func newInt32(val int) *int32 {
pkg/controller/deployment/sync_test.go:func intOrStrP(val int) *intstr.IntOrString {
pkg/kubectl/autoscale_test.go:func newInt32(value int) *int32 {
plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:func userIDPtr(i int) *int64 {
plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:func groupIDPtr(i int) *int64 {
test/e2e/apps/deployment.go:func intOrStrP(num int) *intstr.IntOrString {
test/integration/deployment/util.go:func intOrStrP(num int) *intstr.IntOrString {
```

Here's the helper function from `kubernetes/pkg/apis/core/validation/validation_test.go`:
https://github.com/kubernetes/kubernetes/blob/aa02c0f5191af40cbf7a18a8beec24c56ed7b860/pkg/apis/core/validation/validation_test.go#L1877-L1882

Should we instead look at ensuring that the type is always either `int32` or `int64`?

justaugustus commented:
Both good points! I can update this. :)

justaugustus commented:
@php-coder Updated!
Can you `/approve`, `/lgtm`, `/ok-to-test` to send this to CI?

php-coder commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -12443,17 +12436,17 @@ func TestValidateOrSetClientIPAffinityConfig(t *testing.T) {
 		"non-empty config, valid timeout: core.MaxClientIPServiceAffinitySeconds": {
 			ClientIP: &core.ClientIPConfig{
-				TimeoutSeconds: newInt32(int(core.MaxClientIPServiceAffinitySeconds)),
+				TimeoutSeconds: utilpointer.Int32Ptr(int(core.MaxClientIPServiceAffinitySeconds)),
```
Looks like we don't need to cast to `int` anymore as `MaxClientIPServiceAffinitySeconds` has `int32` type. And above too.

justaugustus replied:
Fixed

php-coder commented on pkg/util/pointer/pointer.go:
```diff
@@ -52,6 +52,12 @@ func Int32Ptr(i int32) *int32 {
+// Int64Ptr returns a pointer to an int64
+func Int64Ptr(i int64) *int64 {
+	o := i
+	return &o
```
Why not simplify it to `return &i` ?

php-coder replied:
Ah, probably, for consistency with `Int32Ptr()`? Ok.

justaugustus replied:
Yeah, this was just for consistency, but I can simplify all of the functions, it if you'd like.

justaugustus replied:
Updated w/ simplified functions.

php-coder commented on plugin/pkg/admission/security/podsecuritypolicy/admission_test.go:
```diff
@@ -1433,7 +1434,7 @@ func TestAdmitFSGroup(t *testing.T) {
 		// admission controller uses DetermineEffectiveSC to get the defaulting
 		// behavior so it can validate what will be applied at runtime
-		pod.Spec.SecurityContext.FSGroup = groupIDPtr(group)
+		pod.Spec.SecurityContext.FSGroup = utilpointer.Int64Ptr(group)
```
Could we change the type of the `group` to `int64`?

justaugustus replied:
Fixed

php-coder commented:
@justaugustus Yes, I can run the tests and put LGTM later.

/ok-to-test

justaugustus commented:
Tests pass!
@php-coder @tallclair @smarterclayton @ericchiang -- can you take a look and approve?

php-coder commented:
/lgtm

@justaugustus If you push the commits again for some other changes, please, extract generated changes to a separate commit. Otherwise, I think, we can leave it as-is.

justaugustus commented:
@php-coder will do!

justaugustus commented:
/test pull-kubernetes-unit

justaugustus commented:
@tallclair PTAL and approve, if all is good

justaugustus commented:
@smarterclayton @liggitt ping for `/approve`

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60073)
<!-- Reviewable:end -->

liggitt commented:
/approve

justaugustus commented:
/retest

justaugustus commented:
/test pull-kubernetes-e2e-gce

justaugustus commented:
/retest

justaugustus committed:
Update bazel rules

justaugustus commented:
@php-coder can you hit this with a `/lgtm`? I just rebased.

php-coder commented:
/lgtm

justaugustus commented:
@smarterclayton could you `/approve` this?

justaugustus commented:
@thockin Could you take a look? It's a quick one; just needs a final `/approve`.
/assign @thockin

thockin commented:
/approve no-issue
